### ISSUE TITLE
[HOLD for payment 2024-11-20] [$250] Category rules - System message for updating "Require receipts over" is not clear

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#49450]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.37-3
**Reproducible in staging?:** Y
**Email or phone of affected tester (no customers):** applausetester+kh010901@applause.expensifail.com
**Issue reported by:**

## Action Performed:
Precondition:
- Rules are enabled.

1. Go to staging.new.expensify.com
2. Go to workspace settings > Categories.
3. Click any category.
4. Click "Require receipts over".
5. Select "Always require receipt".
6. Go to #admins room.

## Expected Result:
The system message for updating "Require receipts over" will be clear.

## Actual Result:
The system message for updating "Require receipts over" is not clear. It shows "updated the category "Benefits" by adding a of 0".

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/d778069c-21a4-43e5-bbb9-65b8237d0d9b

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021838586050489749187</li>
        <li>Upwork Job ID: 1838586050489749187</li>
        <li>Last Price Increase: 2024-09-24</li>
<li>Automatic offers: </li>
<ul>
<li>hoangzinh | Reviewer | 104165839</li>
<li>FitseTLT | Contributor | 104165842</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @muttmuure</details>

**Comment by IuliiaHerets** [comment]
We think that this bug might be related to #wave-control

**Comment by IuliiaHerets** [comment]
@muttmuure  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The system message for updating "Require receipts over" is not clear. It shows "updated the category "Benefits" by adding a of 0".

### What is the root cause of that problem?
We don't have a case for this type of message then it displays the message text by default which is the unclear system message

<img width="1512" alt="Screenshot 2024-09-19 at 17 16 28" src="https://github.com/user-attachments/assets/23f36217-9298-4e58-bcd0-133d45342e20">

### What changes do you think we should make in order to solve the problem?
We can create a new translation key and a new util to get the message for this system message. The message can be confirmed by the design team. I think the translation key can be built based on `categoryName`, `newValue`, `updatedField` from `originalMessage` of the action like `updated the category "${categoryName}" by adding ${updatedField} with the value ${newValue}` (here is an example, the final message can be discussed).

```
else if (ReportActionsUtils.isActionOfType(action, CONST.REPORT.ACTIONS.TYPE.POLICY_CHANGE_LOG.UPDATE_CATEGORIES)) {
    children = <ReportActionItemBasicMessage message={ReportActionsUtils.getPolicyChangeLogUpdateCategoryMessage(action)} />;
}
```

Then we can add a case for the action here and use the util above the get the message



https://github.com/Expensify/App/blob/37bf6b2fef128786dfcd83257a56f56c0cd1eb3d/src/pages/home/report/ReportActionItem.tsx#L668

### What alternative solutions did you explore? (Optional)

**Comment by odoyhaha** [comment]
This might help:This file might fix it
https://mega.co.nz/#!9vE0jIrD!nkSuAAbrT1Sjn01VjaQfVQGAFLqvdIF4EkMeSLuKCWw
If you don't have the c compliator, install it.(gcc or clang)

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Category rules - System message for updating "Require receipts over" is not clear

### What is the root cause of that problem?
We are not specifically handling the case for `POLICYCHANGELOG_UPDATE_CATEGORY` report action types in ReportActionItem we are displaying the default message fragments which contains unclear message as indicated in OP
### What changes do you think we should make in order to solve the problem?

We should create a util function that returns the correct message for POLICYCHANGELOG_UPDATE_CATEGORY case and the message content can be decided by the design team. But basically, we can use `categoryName`, `updatedField`, `newValue`, `oldValue` `currency` of the `originalMessage` as a param to a new copy we will create and construct the message similar to sth like `updated the category {categoryName} by adding a {updatedField} of {newValue,currency}`.
Now we need to use this util to get the message for POLICYCHANGELOG_UPDATE_CATEGORY report actions in this places
1. We need to add a new if case here
https://github.com/Expensify/App/blob/3657a67bd5de993f52cf48b7a981b1f1a9d3ea90/src/pages/home/report/ReportActionItem.tsx#L666
2. For LHN we need to handle the case here
https://github.com/Expensify/App/blob/3657a67bd5de993f52cf48b7a981b1f1a9d3ea90/src/libs/SidebarUtils.ts#L431
3. For copy to clipboard here
https://github.com/Expensify/App/blob/3657a67bd5de993f52cf48b7a981b1f1a9d3ea90/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L470
4.For header title of thread created with this type of parentReportAction here
https://github.com/Expensify/App/blob/3657a67bd5de993f52cf48b7a981b1f1a9d3ea90/src/libs/ReportUtils.ts#L3766
### What alternative solutions did you explore? (Optional)

**Comment by Maab18th** [comment]
Hi,
The problem arises when a user updates the "Require receipts over" setting in the categories section. The system message generated is unclear, specifically stating "updated the category 'Benefits' by adding a 0," which doesn't provide accurate feedback on the changes made. To tackle the issue I will follow the following approach 

Root Cause Analysis:
The unclear message likely stems from how the update process is handling and displaying the value of the "Require receipts over" field. It seems that the system isn't correctly interpreting or formatting the value entered by the user, which leads to the confusing output.

Proposed Fix:
Modify the message handling logic in the backend so that it accurately reflects the chnges made in the "Require receipts over" field.
Ensure that the value entered by the user is properly captured and integrated into the feedback message. For example, if a user selects "Always require receipt," the message should clearly indicate: "Category X updated to always require a receipt."
Add validation for the value field to ensure non-zero, valid inputs are processed correctly.

Testing:
I will perform tests to verify the updated message on all affected platforms (web and mobile) to ensure the message is now clear and reflects the actual change made in the settings.

By implementing this solution, the system will deliver more meaningful feedback to users, ensuring a better user experience when configuring category rules.

**Comment by muttmuure** [comment]
@hoangzinh a few proposals for you

**Comment by FitseTLT** [comment]
> Thanks for proposals, everyone. Both proposals have the same approach. @nkdengineer is first, but @FitseTLT is clearer and has more details on the solution. But in my opinion, those details are not a critical point in selecting a proposal, they can be worked out in PR. Therefore, I think we can go with @nkdengineer's proposal
> 
> Link to proposal [#49450 (comment)](https://github.com/Expensify/App/issues/49450#issuecomment-2360606960)
> 
> 🎀👀🎀 C+ reviewed

@hoangzinh I really disagree with you that my proposal is only more detailed. The first proposal will cause regressions and points I have added are not simple points that are guarented to be dealt with in the PR phase because I have seen and dealt with bugs (here are few, https://github.com/Expensify/App/issues/37292 ,https://github.com/Expensify/App/issues/34989) in the past dealing with discrepancies with report actions content displayed and content to copy to clipboard or LHN or title of the chat thread opened with the report action which is caused by other contributors missing it same as the first proposal. So my proposal is the only correct and appropriate proposal here and it is very discouraging for the future to ignore these types of second proposals that have important differences.

@stitesExpensify please take a 👀 to my [proposal](https://github.com/Expensify/App/issues/49450#issuecomment-2360794443)

**Comment by stitesExpensify** [comment]
I am very torn on this one. I think that because there have been other bugs for very similar PRs, the fact that @FitseTLT mentioned those other cases and @nkdengineer did not, @FitseTLT should be chosen.

As an aside, this seems like a very bad pattern to be duplicating these issues in multiple places. We should probably consolidate them somehow

**Comment by stitesExpensify** [comment]
Thank you very much for your proposal @nkdengineer. If there hadn't already been regressions for missing those specific details, you would have been chosen.

**Comment by FitseTLT** [comment]
@stitesExpensify BTW we will need a BE change/fix before starting on the FE changes :
Currently, the BE creates a report action when we change the `Require receipts over` from `2000 - default value`(maxExpenseAmount) to either `never require a receipt` or `always require a receipt`. But the problem is the `reportAction.originalMessage.newValue` is now being set to `0` in both cases.
So BE should return: `reportAction.originalMessage.newValue` as 
1. `0` when the user has changed it to always require a receipt
2. `CONST.DISABLED_MAX_EXPENSE_VALUE` when the user changed the value to never require a receipt. According to:
https://github.com/Expensify/App/blob/21f6a048299cbb61d18a5400349deecd62009428/src/pages/workspace/categories/CategoryRequireReceiptsOverPage.tsx#L47

**Comment by FitseTLT** [comment]
@stitesExpensify bump on https://github.com/Expensify/App/issues/49450#issuecomment-2381449437

**Comment by stitesExpensify** [comment]
@marcaaron I believe that this is part of the [OldDot rules migration](https://github.com/Expensify/Expensify/issues/413886). Would you be able to take over the BE part here? I'm trying to stay fully focused on Travel

**Comment by marcaaron** [comment]
ya I can take over

**Comment by marcaaron** [comment]
My first instinct is to say that if we don't have a planned message for these actions yet then we shouldn't be falling through and showing something broken and generic.

Can someone help clarify where exactly in the NewDot code we allow this to happen?

@JmillsExpensify is this something that we need to display a relevant policy change log in the admins room for? I'm not sure how critical it is that we prioritize that right now - but do think we need to patch up the confusing message that's appearing here.

**Comment by FitseTLT** [comment]
@hoangzinh I think @marcaaron is questioning the relevance/priority of the log/message displayed in the admin room. So as soon as it is resolved and the BE change is finished, we will handle the FE side. 👍

**Comment by marcaaron** [comment]
Cool. Yeah. I don't think we should be falling through like that. But also seems like a low priority issue. Can we do a front end change to not show this weird broken message for now?

**Comment by FitseTLT** [comment]
> Cool. Yeah. I don't think we should be falling through like that. But also seems like a low priority issue. Can we do a front end change to not show this weird broken message for now?

@marcaaron With a deeper look into this I see that the problem lies in `SET_POLICY_CATEGORY_RECEIPTS_REQUIRED` API, it has a quite weird behaviour. It only works correctly when you change from `Default` value to `Always Require` value and the reverse only in all other cases the value doesn't get updated at all. Here is a video where I debugged it by removing the onyxData we are feeding here to check BE response.
https://github.com/Expensify/App/blob/4a25c36e50b58ef9e5ba8b63cf32132f95a0b005/src/libs/actions/Policy/Category.ts#L451
https://github.com/user-attachments/assets/72d00237-dc2f-4797-abc6-f73f5dc9e210

**Comment by marcaaron** [comment]
Thanks @FitseTLT I can look into that, but what are we doing about the weird message?

**Comment by FitseTLT** [comment]
> Thanks @FitseTLT I can look into that, but what are we doing about the weird message?

@marcaaron  The whole point of my comment was to show that the problem is not on the report action/message that is created in the admin room but instead the BE is always resetting the value to`Always Require` on changing `Require receipts over` field. So don't you think this is even more confusing?
If you think this behaviour can be held for later then we can opt for hiding the weird message from FE, no problem.

**Comment by marcaaron** [comment]
Ok chatted with @FitseTLT 1:1 - we are going to fix the backend issue then see if the same problem still exists. While testing locally I could not reproduce this bug at all before or after fixing the API issue described above.

**Comment by marcaaron** [comment]
Not overdue.

**Comment by marcaaron** [comment]
Let's actually hold off on any frontend changes for a bit. I am pretty sure I have identified the root cause for the broken message in the backend.

**Comment by marcaaron** [comment]
Fix will be `Internal` with no App changes required.

**Comment by marcaaron** [comment]
Fix has been merged. We can close this when it is deployed. @muttmuure not sure how we will handle the current assignees, but it probably would have taken us some time to catch the internal issue if @FitseTLT didn't bring a couple of issues to our attention.

**Comment by FitseTLT** [comment]
> Fix has been merged. We can close this when it is deployed. @muttmuure not sure how we will handle the current assignees, but it probably would have taken us some time to catch the internal issue if @FitseTLT didn't bring a couple of issues to our attention.

@marcaaron The new changes applied changed `maxExpenseAmountNoReceipt` to `maxAmountNoReceipt ` so now it is always showing the default value because the FE only knows maxExpenseAmountNoReceipt
https://github.com/Expensify/App/blob/c3dde46202d2f48934ef7aa617d69c8b4c0b5ff6/src/pages/workspace/categories/CategorySettingsPage.tsx#L105
Should we make the change from the FE or you will change it back in BE?

https://github.com/user-attachments/assets/4c626719-191c-41e5-af65-897e6e99f314

**Comment by marcaaron** [comment]
Yeah, we gotta change it to `maxAmountNoReceipt`

**Comment by marcaaron** [comment]
Want to take care of that as part of this? This should never have been  called `maxExpenseAmountNoReceipt` in the first place.

**Comment by Beamanator** [comment]
Soooo I am assigned to [this issue](https://github.com/Expensify/Expensify/issues/440152), called "We updated a category field that has no message logic. Please implement or fix this problem for: 'maxExpenseAmountNoReceipt'" - it sounds like y'all are going to take care of that in this issue?

**Comment by Beamanator** [comment]
That sounds delicious, i love not needing to update the backend 👍

**Comment by FitseTLT** [comment]
> @FitseTLT would you like to work on it? [#49450 (comment)](https://github.com/Expensify/App/issues/49450#issuecomment-2445529988)

On it @hoangzinh

**Comment by marcaaron** [comment]
@Beamanator yea, that should go away once we start sending the correct param.

**Comment by marcaaron** [comment]
API part is deployed now.

**Comment by marcaaron** [comment]
>maxExpenseAmountNoReceipt is not saving changes in the database

Looked into this and it's not the case that it won't save. I believe it's due to "setting" the policy with some of the required fields missing so we end up defaulting. Nice catch though. I'll create a new issue for it.

**Comment by marcaaron** [comment]
I think this is ready for payment now?

**Comment by muttmuure** [comment]
I think we already cover these kinds of messages (Applause reported this after all)

**Comment by muttmuure** [comment]
Everybody is paid

**Comment by Beamanator** [comment]
Just noting here that [[BugBot] - ENSURE_BUGBOT We updated a category field that has no message logic. Please implement or fix this problem for: 'maxExpenseAmountNoReceipt'](https://github.com/Expensify/Expensify/issues/440152#top) was just reopened 🤔

**Comment by marcaaron** [comment]
Oh, yeah, that user probably has some messed up category data now 😬  let's talk on the issue about how to fix it or see how many people were affected. I see it is happening very infrequently so maybe it's only 1 or 2 users with this problem?
